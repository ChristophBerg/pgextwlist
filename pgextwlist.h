/* PostgreSQL Extension WhiteList -- Dimitri Fontaine
 *
 * Author: Dimitri Fontaine <dimitri@2ndQuadrant.fr>
 * Licence: PostgreSQL
 * Copyright Dimitri Fontaine, 2011-2013
 *
 * For a description of the features see the README.md file from the same
 * distribution.
 */

#ifdef PG_VERSION_NUM
#define PG_MAJOR_VERSION (PG_VERSION_NUM / 100)
#else
#error "Unknown PostgreSQL version"
#endif

#if PG_MAJOR_VERSION != 901    \
	&& PG_MAJOR_VERSION != 902 \
	&& PG_MAJOR_VERSION != 903 \
	&& PG_MAJOR_VERSION != 904
#error "Unsupported postgresql version"
#endif

