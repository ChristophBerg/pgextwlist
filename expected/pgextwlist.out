CREATE ROLE mere_mortal;
SET ROLE mere_mortal;
SHOW extwlist.extensions;
     extwlist.extensions      
------------------------------
 citext,earthdistance,pg_trgm
(1 row)

SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 plpgsql | 1.0
(1 row)

-- pre-existing extension
CREATE EXTENSION plpgsql;
ERROR:  extension "plpgsql" already exists
SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 plpgsql | 1.0
(1 row)

-- non-whitelisted extension
CREATE EXTENSION hstore;
ERROR:  permission denied to create extension "hstore"
HINT:  Must be superuser to create this extension.
SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 plpgsql | 1.0
(1 row)

-- whitelisted extensions
CREATE EXTENSION citext;
CREATE EXTENSION pg_trgm;
SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 citext  | 1.0
 pg_trgm | 1.1
 plpgsql | 1.0
(3 rows)

-- whitelisted extension, but dependency is missing
CREATE EXTENSION earthdistance;
ERROR:  required extension "cube" is not installed
SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 citext  | 1.0
 pg_trgm | 1.1
 plpgsql | 1.0
(3 rows)

-- drop whitelisted extension
DROP EXTENSION pg_trgm;
SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 citext  | 1.0
 plpgsql | 1.0
(2 rows)

-- drop non-whitelisted extension
DROP EXTENSION plpgsql;
ERROR:  must be owner of extension plpgsql
SELECT extname, extversion FROM pg_extension ORDER BY 1;
 extname | extversion 
---------+------------
 citext  | 1.0
 plpgsql | 1.0
(2 rows)

